### ISSUE TITLE
free-threaded x86-64 macOS wheel build job failed with pybind11 2.13.0

### ISSUE THREAD

**Issue by rgommers** [scipy/scipy#21056]
From the [CI log](https://github.com/scipy/scipy/actions/runs/9676956383/job/26697644175) from an hour ago:
```
         pybind11-config found: YES (/private/var/folders/4z/th_0342d7sv9wl7kb3yfllq80000gn/T/cibw-run-ssqqra29/cp313t-macosx_x86_64/build/venv/bin/pybind11-config) 2.13.0
        Run-time dependency pybind11 found: YES 2.13.0
...

        [770/1386] Compiling C++ object scipy/fft/_pocketfft/pypocketfft.cpython-313t-darwin.so.p/pypocketfft.cxx.o
        FAILED: scipy/fft/_pocketfft/pypocketfft.cpython-313t-darwin.so.p/pypocketfft.cxx.o
        c++ -Iscipy/fft/_pocketfft/pypocketfft.cpython-313t-darwin.so.p -Iscipy/fft/_pocketfft -I../scipy/fft/_pocketfft -I../scipy/_lib/pocketfft -I/Library/Frameworks/PythonT.framework/Versions/3.13/include/python3.13t -I/private/var/folders/4z/th_0342d7sv9wl7kb3yfllq80000gn/T/cibw-run-ssqqra29/cp313t-macosx_x86_64/build/venv/lib/python3.13/site-packages/pybind11/include -fvisibility=hidden -fvisibility-inlines-hidden -fdiagnostics-color=always -DNDEBUG -Wall -Winvalid-pch -std=c++17 -O3 -DPOCKETFFT_PTHREADS -DPOCKETFFT_CACHE_SIZE=16 -MD -MQ scipy/fft/_pocketfft/pypocketfft.cpython-313t-darwin.so.p/pypocketfft.cxx.o -MF scipy/fft/_pocketfft/pypocketfft.cpython-313t-darwin.so.p/pypocketfft.cxx.o.d -o scipy/fft/_pocketfft/pypocketfft.cpython-313t-darwin.so.p/pypocketfft.cxx.o -c ../scipy/fft/_pocketfft/pypocketfft.cxx
        In file included from ../scipy/fft/_pocketfft/pypocketfft.cxx:15:
        In file included from /private/var/folders/4z/th_0342d7sv9wl7kb3yfllq80000gn/T/cibw-run-ssqqra29/cp313t-macosx_x86_64/build/venv/lib/python3.13/site-packages/pybind11/include/pybind11/pybind11.h:13:
        In file included from /private/var/folders/4z/th_0342d7sv9wl7kb3yfllq80000gn/T/cibw-run-ssqqra29/cp313t-macosx_x86_64/build/venv/lib/python3.13/site-packages/pybind11/include/pybind11/detail/class.h:12:
        In file included from /private/var/folders/4z/th_0342d7sv9wl7kb3yfllq80000gn/T/cibw-run-ssqqra29/cp313t-macosx_x86_64/build/venv/lib/python3.13/site-packages/pybind11/include/pybind11/detail/../attr.h:14:
        In file included from /private/var/folders/4z/th_0342d7sv9wl7kb3yfllq80000gn/T/cibw-run-ssqqra29/cp313t-macosx_x86_64/build/venv/lib/python3.13/site-packages/pybind11/include/pybind11/cast.h:15:
        In file included from /private/var/folders/4z/th_0342d7sv9wl7kb3yfllq80000gn/T/cibw-run-ssqqra29/cp313t-macosx_x86_64/build/venv/lib/python3.13/site-packages/pybind11/include/pybind11/detail/type_caster_base.h:15:
        /private/var/folders/4z/th_0342d7sv9wl7kb3yfllq80000gn/T/cibw-run-ssqqra29/cp313t-macosx_x86_64/build/venv/lib/python3.13/site-packages/pybind11/include/pybind11/detail/internals.h:557:46: error: aligned allocation function of type 'void *(std::size_t, std::align_val_t)' is only available on macOS 10.14 or newer
                internals_ptr->instance_shards.reset(new instance_map_shard[num_shards]);
                                                     ^
        /private/var/folders/4z/th_0342d7sv9wl7kb3yfllq80000gn/T/cibw-run-ssqqra29/cp313t-macosx_x86_64/build/venv/lib/python3.13/site-packages/pybind11/include/pybind11/detail/internals.h:557:46: note: if you supply your own aligned allocation functions, use -faligned-allocation to silence this diagnostic
        /private/var/folders/4z/th_0342d7sv9wl7kb3yfllq80000gn/T/cibw-run-ssqqra29/cp313t-macosx_x86_64/build/venv/lib/python3.13/site-packages/pybind11/include/pybind11/detail/internals.h:557:46: error: aligned deallocation function of type 'void (void *, std::align_val_t) noexcept' is only available on macOS 10.14 or newer
                internals_ptr->instance_shards.reset(new instance_map_shard[num_shards]);
                                                     ^
        /private/var/folders/4z/th_0342d7sv9wl7kb3yfllq80000gn/T/cibw-run-ssqqra29/cp313t-macosx_x86_64/build/venv/lib/python3.13/site-packages/pybind11/include/pybind11/detail/internals.h:557:46: note: if you supply your own aligned allocation functions, use -faligned-allocation to silence this diagnostic
        In file included from ../scipy/fft/_pocketfft/pypocketfft.cxx:15:
        In file included from /private/var/folders/4z/th_0342d7sv9wl7kb3yfllq80000gn/T/cibw-run-ssqqra29/cp313t-macosx_x86_64/build/venv/lib/python3.13/site-packages/pybind11/include/pybind11/pybind11.h:13:
        In file included from /private/var/folders/4z/th_0342d7sv9wl7kb3yfllq80000gn/T/cibw-run-ssqqra29/cp313t-macosx_x86_64/build/venv/lib/python3.13/site-packages/pybind11/include/pybind11/detail/class.h:12:
        In file included from /private/var/folders/4z/th_0342d7sv9wl7kb3yfllq80000gn/T/cibw-run-ssqqra29/cp313t-macosx_x86_64/build/venv/lib/python3.13/site-packages/pybind11/include/pybind11/attr.h:13:
        In file included from /private/var/folders/4z/th_0342d7sv9wl7kb3yfllq80000gn/T/cibw-run-ssqqra29/cp313t-macosx_x86_64/build/venv/lib/python3.13/site-packages/pybind11/include/pybind11/detail/common.h:317:
        In file included from /Applications/Xcode_13.4.1.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.sdk/usr/include/c++/v1/forward_list:184:
        In file included from /Applications/Xcode_13.4.1.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.sdk/usr/include/c++/v1/algorithm:653:
        In file included from /Applications/Xcode_13.4.1.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.sdk/usr/include/c++/v1/functional:500:
        In file included from /Applications/Xcode_13.4.1.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.sdk/usr/include/c++/v1/__functional/function.h:20:
        In file included from /Applications/Xcode_13.4.1.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.sdk/usr/include/c++/v1/__memory/shared_ptr.h:25:
        /Applications/Xcode_13.4.1.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.sdk/usr/include/c++/v1/__memory/unique_ptr.h:88:5: error: aligned deallocation function of type 'void (void *, std::align_val_t) noexcept' is only available on macOS 10.14 or newer
            delete[] __ptr;
            ^
        /Applications/Xcode_13.4.1.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.sdk/usr/include/c++/v1/__memory/unique_ptr.h:551:7: note: in instantiation of function template specialization 'std::default_delete<pybind11::detail::instance_map_shard []>::operator()<pybind11::detail::instance_map_shard>' requested here
              __ptr_.second()(__tmp);
              ^
        /Applications/Xcode_13.4.1.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.sdk/usr/include/c++/v1/__memory/unique_ptr.h:495:19: note: in instantiation of member function 'std::unique_ptr<pybind11::detail::instance_map_shard []>::reset' requested here
          ~unique_ptr() { reset(); }
                          ^
        /private/var/folders/4z/th_0342d7sv9wl7kb3yfllq80000gn/T/cibw-run-ssqqra29/cp313t-macosx_x86_64/build/venv/lib/python3.13/site-packages/pybind11/include/pybind11/detail/internals.h:212:5: note: in instantiation of member function 'std::unique_ptr<pybind11::detail::instance_map_shard []>::~unique_ptr' requested here
            ~internals() {
            ^
        /Applications/Xcode_13.4.1.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.sdk/usr/include/c++/v1/__memory/unique_ptr.h:88:5: note: if you supply your own aligned allocation functions, use -faligned-allocation to silence this diagnostic
            delete[] __ptr;
            ^
        3 errors generated.
        [771/1386] Generating 'scipy/stats/_rcont/rcont.cpython-313t-darwin.so.p/rcont.c'
        [772/1386] Linking target scipy/io/_test_fortran.cpython-313t-darwin.so
        [773/1386] Generating 'scipy/stats/_unuran/unuran_wrapper.cpython-313t-darwin.so.p/unuran_wrapper.c'
        [774/1386] Generating 'scipy/stats/_stats_pythran.cpython-313t-darwin.so.p/_stats_pythran.cpp'
        <frozen importlib._bootstrap>:488: RuntimeWarning: The global interpreter lock (GIL) has been enabled to load module 'numpy._core._multiarray_umath', which has not declared that it can run safely without the GIL. To override this behavior and keep the GIL disabled (at your own risk), run with PYTHON_GIL=0 or -Xgil=0.
        [775/1386] Generating 'scipy/io/matlab/_streams.cpython-313t-darwin.so.p/_streams.c'
        ninja: build stopped: subcommand failed.
        [end of output]
```

`pybind11` 2.13.0 was released earlier today and is the first version that should have free-threaded CPython support. So the changes are probably not interacting too well with something in `pypocketfft`.

**Comment by rgommers** [comment]
For ease of reading, the error is:
```
include/pybind11/detail/internals.h:557:46: error: aligned allocation function of type 'void *(std::size_t, std::align_val_t)' is only available on macOS 10.14 or newer
```

It's likely coming from https://github.com/pybind/pybind11/pull/5148, but not intentionally. There is only one mention of alignment in that PR, and one resolved comment about a bump in `PYBIND11_INTERNALS_VERSION` for the free-threaded build. So without investigating in much detail, I suspect that it may be an existing check for macOS <10.14 that is now being triggered.

Would be good to understand this; if needed we can bump the deployment target to 10.14 or 10.15, that shouldn't affect anyone in the real world.

Cc @colesbury @henryiii for visibility -in case you happen to know what the root cause is here based on the traceback above, we'd of course be keen to hear that.

**Comment by lucascolley** [comment]
aligned allocation is the usual pocketfft suspect (https://github.com/mreineck/pocketfft/pull/11)

**Comment by colesbury** [comment]
Ooops, I forgot `alignas(64)` causes problems.

**Comment by colesbury** [comment]
https://github.com/pybind/pybind11/blob/895e657220c1ff2058a4cfee6b185bd0582b0157/include/pybind11/detail/internals.h#L155-L159

**Comment by henryiii** [comment]
Related but not really: I see `std::mutex` is here, since PyMutex has become public, I think we should try moving to that, as well.

**Comment by rgommers** [comment]
Confirmed that setting `MACOSX_DEPLOYMENT_TARGET` to 10.14 fixes the problem. There may be other problems though, and having everyone bump their deployment target is also not ideal. Discussed with @colesbury just now: he'd like to change this in `pybind11`.

**Comment by colesbury** [comment]
PR in pybind11: https://github.com/pybind/pybind11/pull/5200

**Comment by henryiii** [comment]
Missed by exactly one macOS version (3.13 requires 10.13+). 🤦

**Comment by rgommers** [comment]
> Missed by exactly one macOS version (3.13 requires 10.13+).

Ah, thanks for pointing that out. CI config here still sets `10.9`, but `cibuildwheel` then produces `scipy-1.15.0.dev0-cp313-cp313t-macosx_10_13_x86_64.whl` indeed. And we upgraded to C++17, so according to https://cibuildwheel.pypa.io/en/stable/cpp_standards/#macos-and-deployment-target-versions we should be using `10.13` everywhere.

**Comment by henryiii** [comment]
2.13.1 is out.

**Comment by rgommers** [comment]
@mreineck thanks for having a look. It's not pocketfft itself this time, it's simply the first place the build hits where we used `pybind11` with vendored C++ code. `alignas` was already removed from `pybind11`, so the problem is solved with pybind11 2.13.1.

Thanks for the quick fix @colesbury and @henryiii!
